terraform/modules/example/main.tf

```diff
@@ -155,34 +177,45 @@ resource "aws_sns_topic" "main" {
 # Lifecycle hook
 resource "aws_autoscaling_lifecycle_hook" "main" {
   name                    = "${var.name_prefix}-lifecycle"
-  autoscaling_group_name  = "${module.asg.id}"
+  autoscaling_group_name  = "${aws_autoscaling_group.main.id}"
   lifecycle_transition    = "autoscaling:EC2_INSTANCE_TERMINATING"
   default_result          = "CONTINUE"
   heartbeat_timeout       = "60"
   notification_target_arn = "${aws_sns_topic.main.arn}"
   role_arn                = "${aws_iam_role.lifecycle.arn}"
 }
 
+resource "aws_iam_instance_profile" "lifecycle" {
+  name = "${var.name_prefix}-lifecycle-instance-profile"
+  role = "${aws_iam_role.lifecycle.name}"
+}
+
 # Execution role and policies for the lifecycle hook
 resource "aws_iam_role" "lifecycle" {
   name               = "${var.name_prefix}-lifecycle-role"
-  assume_role_policy = "${data.aws_iam_policy_document.asg_assume.json}"
+  assume_role_policy = "${data.aws_iam_policy_document.assume.json}"
+}
+
+resource "aws_iam_role_policy" "lifecycle-asg" {
+  name   = "${var.name_prefix}-lifecycle-asg-permissions"
+  role   = "${aws_iam_role.lifecycle.id}"
+  policy = "${data.aws_iam_policy_document.asg_permissions.json}"
 }
 
 resource "aws_iam_role_policy" "lifecycle" {
   name   = "${var.name_prefix}-lifecycle-permissions"
   role   = "${aws_iam_role.lifecycle.id}"
-  policy = "${data.aws_iam_policy_document.asg_permissions.json}"
+  policy = "${data.aws_iam_policy_document.permissions.json}"
 }
 
-data "aws_iam_policy_document" "asg_assume" {
+data "aws_iam_policy_document" "assume" {
   statement {
     effect  = "Allow"
     actions = ["sts:AssumeRole"]
 
     principals {
       type        = "Service"
-      identifiers = ["autoscaling.amazonaws.com"]
+      identifiers = ["ec2.amazonaws.com", "autoscaling.amazonaws.com"]
```

Comment:
I think it's better to use two different roles so that the example policies can (more or less) be copy/pasted into a production deployment. Right now we are granting `sns:Publish` to the autoscaling group, which it does not need, and a lot of extra permissions to the lifecycle hook which it also does not need.